labscript_devices/PrawnDO/labscript_devices.py

```diff
@@ -0,0 +1,292 @@
+                 clock_frequency = 100e6,
+                 external_clock = False,
+                ):
+        """PrawnDO digital output device.
+        
+        This labscript device provides general purpose digital outputs
+        using a Raspberry Pi Pico with custom firmware.
+
+        It supports two types of connections to a parent device:
+        direct to a :class:`~.Clockline` via the `clock_line` argument or
+        through a :class:`~.Trigger` from an :class:`~.IntermediateDevice`
+        via the `trigger_device` and `trigger_connection` arguments.
+        Only one should be supplied.
+
+
+        Args:
+            name (str): python variable name to assign to the PrawnDO
+            trigger_device (:class:`~.IntermediateDevice`, optional):
+                Device that will send the starting hardware trigger.
+                Used when connecting to an `IntermediateDevice` via a `DigitalOut`.
+            trigger_connection (str, optional): Which output of the `trigger_device`
+                is connected to the PrawnDO hardware trigger input.
+                Not required when directly connected to a `Clockline`.
+            clock_line (:class:`~.Clockline`, optional):
+                Used when connected directly to a `Clockline`.
+                Not required if using a trigger device.
+            com_port (str): COM port assinged to the PrawnDO by the OS.
+                Takes the form of `COMd` where `d` is an integer.
+            clock_frequency (float, optional): System clock frequency, in Hz.
+                Must be less than 133 MHz. Default is `100e6`.
+            external_clock (bool, optional): Whether to use an external clock.
+                Default is `False`.
+        """
+
+        if clock_frequency > 133e6:
+            raise ValueError('Clock frequency must be less than 133 MHz')
+        
+        self.external_clock = external_clock
+        self.clock_frequency = clock_frequency
+        # update specs based on clock frequency
+        if self.clock_frequency != 100e6:
+            # factor to scale times by
+            factor = 100e6/self.clock_frequency
+            self.clock_limit *= factor
+            self.clock_resolution *= factor
+            self.minimum_duration *= factor
+            self.wait_delay *= factor
+            self.input_response_time *= factor
+            self.trigger_delay *= factor
+            self.trigger_minimum_duration *= factor
+
+        if clock_line is not None and trigger_device is not None:
+            raise LabscriptError("Provide only a trigger_device or a clock_line, not both")
+        if clock_line is not None:
+            # make internal Intermediate device and trigger to connect it
+            self.__intermediate = _PrawnDOIntermediateDevice(f'{name:s}__intermediate',
+                                                             clock_line)
+            PseudoclockDevice.__init__(self, name, self.__intermediate, 'internal')
```

Comment:
When running in this mode we need to:
- Effectively disable `PrawnDO.set_initial_trigger_time()` because the intermediate device isn't a real device that will ignore other pseudoclock pulses
- Prevent anyone from manually setting `PrawnDO.initial_trigger_time` (should stay at the default of `0`). Maybe that can be done by defining it as a property with only a getter in this class? E.g.:
  ```
  def __init__(self, ...):
      self._initial_trigger_time = 0
      ...

  @property
  def initial_trigger_time(self):
      return self._initial_trigger_time

  @initial_trigger_time.setter
  def initial_trigger_time(self, value):
      if value != 0 and hasattr(self, "__intermediate"):
          raise LabscriptError("You cannot set the initial trigger time when the PrawnDO is directly triggered by a clockline")
      self._initial_trigger_time = value

  def set_initial_trigger_time(self, *args, **kwargs):
      if hasattr(self, "__intermediate"):
          LabscriptError("You cannot set the initial trigger time when the PrawnDO is directly triggered by a clockline")
      return super().set_initial_trigger_time(*args, **kwargs)
  ```

Reply:
I have incorporated this.